File: python/interpreter/interpreter.py
from abc import ABC


class Expression(ABC):
    def interpret(self, s: str) -> bool:
        raise NotImplementedError()


class Digit(Expression):
    def interpret(self, s: str) -> bool:
        return s in [str(i) for i in range(10)]


class Number(Expression):
    def interpret(self, s: str) -> bool:
        if not s:
            return False
        return Digit().interpret(s) or Digit().interpret(s[0]) and Number().interpret(s[1:])


def main():
    num: Number = Number()
    print(num.interpret("10"))
    print(num.interpret(""))
    print(num.interpret("f10"))
    print(num.interpret("10f"))


if __name__ == '__main__':
    main()

File: cpp/interpreter/interpreter.cpp
#include <iostream>

class Expression {
public:
    virtual bool interpret(const std::string &str) = 0;
};

class Digit : public Expression {
public:
    bool interpret(const std::string &str) override {
        if (str == "0") {
            return true;
        }
        if (str == "1") {
            return true;
        }
        if (str == "2") {
            return true;
        }
        if (str == "3") {
            return true;
        }
        if (str == "4") {
            return true;
        }
        if (str == "5") {
            return true;
        }
        if (str == "6") {
            return true;
        }
        if (str == "7") {
            return true;
        }
        if (str == "8") {
            return true;
        }
        if (str == "9") {
            return true;
        }
        return false;
    }
};

class Number : public Expression {
public:
    bool interpret(const std::string &str) override {
        if (str.empty()) {
            return false;
        }
        std::string first_digit(1, str[0]);
        std::string next_digits(str.substr(1));
        return Digit().interpret(str) || Digit().interpret(first_digit) && Number().interpret(next_digits);
    }
};

int main() {
    Number number;
    std::cout << number.interpret("101") << std::endl;
    std::cout << number.interpret("") << std::endl;
    std::cout << number.interpret("1f1") << std::endl;
    std::cout << number.interpret("101f") << std::endl;
    return 0;
}
